File: src/main/scala/com/lcy/scala/demo/rdd/instance/PageRank.scala
package com.lcy.scala.demo.rdd.instance

import org.apache.spark.{HashPartitioner, SparkConf, SparkContext}

object PageRank {

  def main(args: Array[String]): Unit = {

    val conf = new SparkConf().setAppName("SparkStudy").setMaster("local")
    val sc = new SparkContext(conf)

    val links = sc.parallelize(List(
      ("A",List("A","B","C","D")),
      ("B",List("A","B","C")),
      ("C",List("A","B")),
      ("D",List("A")))
    ).partitionBy(new HashPartitioner(100))
      .persist()

    var ranks = links.mapValues(v=>1.0)

    for (i <- 0 until 10) {
      val contributions = links.join(ranks).flatMap { //flapmap结果("A",(List("B","C"),1))
        case (pageId,(links,rank)) => links.map(dest=>(dest,rank/links.size))
      }
      ranks=contributions.reduceByKey(_+_).mapValues(0.15+0.85*_)
    }

      ranks.sortBy(e => e._2, false).collect().foreach(println)
  }

}

File: src/main/java/com/lcy/java/spark/rdd/instance/PageRank.java
package com.lcy.java.spark.rdd.instance;

import com.google.common.collect.Iterables;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.function.Function2;
import org.apache.spark.sql.SparkSession;
import scala.Tuple2;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Pattern;

public final class PageRank {

    private static final Pattern SPACES = Pattern.compile("\\s+");

    private static class Sum implements Function2<Double, Double, Double> {
        @Override
        public Double call(Double a, Double b) {
            return a + b;
        }
    }

    public static void main(String[] args) throws Exception {

        SparkSession spark = SparkSession
                .builder()
                .appName("JavaPageRank")
                .master("local")
                .getOrCreate();

        JavaRDD<String> lines = spark.read().textFile("file:/D:/sparkData/page.txt").javaRDD();

        JavaPairRDD<String, Iterable<String>> links = lines.mapToPair(s -> {
            String[] parts = SPACES.split(s);
//            String[] parts = s.split(" ");
            return new Tuple2<>(parts[0], parts[1]);
        }).distinct().groupByKey().cache();

        JavaPairRDD<String, Double> ranks = links.mapValues(rs -> 1.0);

        for (int current = 0; current < 10; current++) {
            JavaPairRDD<String, Double> contribs = links.join(ranks).values()
                    .flatMapToPair(s -> {
                        int urlCount = Iterables.size(s._1());
                        List<Tuple2<String, Double>> results = new ArrayList<>();
                        for (String n : s._1) {
                            results.add(new Tuple2<>(n, s._2() / urlCount));
                        }
                        return results.iterator();
                    });

            ranks = contribs.reduceByKey(new Sum()).mapValues(sum -> 0.15 + sum * 0.85);
        }

        List<Tuple2<String, Double>> output = ranks.collect();
        for (Tuple2<?,?> tuple : output) {
            System.out.println(tuple._1() + " has rank: " + tuple._2() + ".");
        }

        spark.stop();
    }
}
